pkg/config/spec.go

```diff
@@ -545,6 +545,47 @@ func ReadUkiUpgradeFromConfig(c *Config) (*v1.UpgradeUkiSpec, error) {
 	return upgradeSpec, nil
 }
 
+// getDirSize will walk through a dir and calculate the value of each file, and will continue doing so until it has visited all files.
+// fileList: keeps track of the files visited to avoid counting a file more than once if it's a symlink. It could also be used as a way to filter some files
+// size: will be the memory that adds up all the files sizes. Meaning it could be initialized with a value greater than 0 if needed.
+func getDirSize(size *int64, fileList map[string]bool, path string, d fs.DirEntry, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if d.IsDir() {
+		return nil
+	}
+
+	actualFilePath := path
+	if d.Type()&fs.ModeSymlink != 0 {
+		// If it's a symlink, get its target and calculate its size.
+		var err error
+		actualFilePath, err = os.Readlink(path)
+		if err != nil {
+			return err
+		}
+
+		if !filepath.IsAbs(actualFilePath) {
+			// If it's a relative path, join it with the base directory path.
+			actualFilePath = filepath.Join(filepath.Dir(path), actualFilePath)
+		}
+	}
+
+	fileInfo, err := os.Stat(actualFilePath)
+	if os.IsNotExist(err) || fileList[actualFilePath] {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	*size += fileInfo.Size()
+	fileList[actualFilePath] = true
```

Comment:
isnt this gonna be terrible for resource compsumption? Like we are really storing a full map of files with a bool? How does this look with a fs with 50000 files?

Reply:
@Itxaka good question. So far, running it on an installation, does not seem to impact the speed, but I'm going to add some debugging to get an idea of how many files we are talking about, and I could use that number to initialize the map so that at least resizing does not need to happen as often

Reply:
mostly thinking about things like rpi with smaller amounts of mem

Reply:
Maybe this helps:

```
package main

import (
	"fmt"
	"runtime"
	"strconv"
)

func main() {
	var mymap = map[string]bool{}

	for i := 0; true; i++ {
		mymap[strconv.Itoa(i)] = true
		if i%100 == 0 {
			fmt.Printf("i = %+v\n", i)
			PrintMemUsage()
		}
	}
}

func PrintMemUsage() {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	fmt.Printf("Alloc = %v MiB", m.Alloc/1024/1024)
	fmt.Printf("\tTotalAlloc = %v MiB", m.TotalAlloc/1024/1024)
	fmt.Printf("\tSys = %v MiB", m.Sys/1024/1024)
	fmt.Printf("\tNumGC = %v\n", m.NumGC)
}

```

Reply:
PrintMemUsage copied from here: https://gophercoding.com/print-current-memory-usage/

Reply:
sample line:

```
i = 13661300
Alloc = 1121 MiB	TotalAlloc = 1493 MiB	Sys = 1361 MiB	NumGC = 9
```

13.5 million keys -> 1.5 Gb RAM ?

Reply:
> sample line:
> 
> ```
> i = 13661300
> Alloc = 1121 MiB	TotalAlloc = 1493 MiB	Sys = 1361 MiB	NumGC = 9
> ```
> 
> 13.5 million keys -> 1.5 Gb RAM ?

so extrapolating from there, 1 million entries = 150Mb, 100k entries 15Mb ?

If that is accurate then it should be negligent to store 50k entries

Reply:
On Ubuntu I see 27K files but I don't see an impact in memory usage at all:

`var filesVisited map[string]bool`
Alloc = 4 MiB
TotalAlloc = 8 MiB

`filesVisited = make(map[string]bool, 30000)`
Alloc = 4 MiB
TotalAlloc = 8 MiB

After appending all the values to the map
Alloc = 4 MiB
TotalAlloc = 8 MiB